### ISSUE TITLE
Stacked Bar Chart Question

### ISSUE THREAD

**Issue by marsvolta80** [dc-js/dc.js#1038]
I have json data that is structured as follows (see below). I'm displaying the data on a 
stacked bar chart where each set of stacked bars represents one day. I'm stacking bars on each day using the source and it's corresponding count:

-- This would stack 6 bars 10/29
{"date": "2015-10-29", "source": "a", "count": 1}
{"date": "2015-10-29", "source": "b", "count": 6} 
{"date": "2015-10-29", "source": "c", "count": 5} 
{"date": "2015-10-29", "source": "d", "count": 3}
{"date": "2015-10-29", "source": "e", "count": 2} 
{"date": "2015-10-29", "source": "f", "count": 0}

-- This would stack 6 bars for 10/30
{"date": "2015-10-30", "source": "a", "count": 1}
{"date": "2015-10-30", "source": "b", "count": 1} 
{"date": "2015-10-30", "source": "c", "count": 2}
{"date": "2015-10-30", "source": "d", "count": 3}
{"date": "2015-10-30", "source": "e", "count": 2}
{"date": "2015-10-30", "source": "f", "count": 4}

-- I was expecting this to stack 3 bars for 10/31 but that didn't happen
{"date": "2015-10-31", "source": "a", "count": 1}
{"date": "2015-10-31", "source": "b", "count": 6} 
{"date": "2015-10-31", "source": "c", "count": 5}

The problem I'm running into is that for 10/31 it's only displaying one bar corresponding to the first entry for 10/31. The bars for b and c don't appear. Through testing I found that it works if I provide the same number of entries per day. So it works fine if I provide zeros for the sources that are missing such as:

{"date": "2015-10-31", "source": "a", "count": 1} 
{"date": "2015-10-31", "source": "b", "count": 6} 
{"date": "2015-10-31", "source": "c", "count": 5}
{"date": "2015-10-30", "source": "d", "count": 0} 
{"date": "2015-10-30", "source": "e", "count": 0}
{"date": "2015-10-30", "source": "f", "count": 0}

I'm wondering is it required for the json to be consistent across the date dimension (I'm using crossfilter to create a date dimension)? In other words is the stacked bar graph expecting to see the same number of source entries per day or does it handle this kind of data inconsistency? I apologize in advance if this question has been answered already.

**Comment by gordonwoodhull** [comment]
Yes, currently the stack mixin requires all stacks to have the same keys. I don't think this requirement is documented, so I think I'll file this as a documentation bug. (It's been covered on SO once or twice.)

One workaround is to combine the groups using new "fake groups":

https://github.com/dc-js/dc.js/wiki/FAQ#combine-groups

Please let me know if this works for you.

One could argue this should be the default behavior, and I think when we have a better data (pre)processing model, this should be part of it.

**Comment by marsvolta80** [comment]
Thanks for the quick reply! I'll give this a shot and let you know.

**Comment by marsvolta80** [comment]
Unless I'm not understanding something (which is definitely possible) I don't think combining groups will work for me because I only have one group to work with. I'll look into this a little more and reply if I get any further. I have a few other work arounds to ensure the json structure is compliant so I'll most likely take that route. Again, thanks for the response!

**Comment by gordonwoodhull** [comment]
Okay, I don't think I understand how you have structured your code. If you have a single group, is it reducing multiple values? Do you mean that it is simply leaving some of the fields/stacks undefined? 

If you're using a custom reduction, you could have the initialization (third) function start all fields/stacks at zero, instead of changing your input data. But again, I'm not sure how you've structured your code.

**Comment by marsvolta80** [comment]
Hey after reading your last comment it sparked some new thoughts for me and I think I got it now! So yes I'm using a custom reduction date.dimension.group().reduce(...). It has the three functions add, subtract, initialize. The problem was with my third function. It was not initializing all of my sources to zero. From what I gather without this the input data is inconsistent. So this was forcing me to make sure I had zero entries in my input data. 

Overall I think I didn't quite grasp the purpose of that third function until now. So what I'm doing now is looking at the sources I receive and just creating a new object that sets each of my sources to zero and is returned from the initialize function. This ensures each source has at least a value of zero. This seems to be working very well. I wanted to do some more testing but so far it looks promising!! Thanks very much for your second response. I don't think this would have occurred to me without you providing that information! Let me know if you think sounds like the right approach.

**Comment by gordonwoodhull** [comment]
Yes, that sounds right to me. We should mention this in the section on multiple-value reductions as well. 

https://github.com/dc-js/dc.js/wiki/FAQ#how-do-i-reduce-multiple-values-at-once

Thanks for confirming!
